core/src/avm2/dynamic_map.rs

```diff
@@ -0,0 +1,215 @@
+use fnv::FnvBuildHasher;
+use gc_arena::Collect;
+use hashbrown::{self, raw::RawTable};
+use std::hash::Hash;
+
+use super::{string::AvmString, Object};
+
+#[derive(Debug, Collect, Copy, Clone)]
+#[collect(no_drop)]
+pub struct DynamicProperty<V> {
+    pub value: V,
+    pub enumerable: bool,
+}
+
+#[derive(Eq, PartialEq, Hash, Copy, Clone, Collect)]
+#[collect(no_drop)]
+pub enum StringOrObject<'gc> {
+    String(AvmString<'gc>),
+    Object(Object<'gc>),
+}
+
+/// A HashMap designed for dynamic properties on an object.
+#[derive(Debug, Collect, Clone)]
+#[collect(no_drop)]
+pub struct DynamicMap<K: Eq + PartialEq + Hash, V> {
+    values: hashbrown::HashMap<K, DynamicProperty<V>, FnvBuildHasher>,
+    // The last index that was given back to flash
+    public_index: usize,
+    // The actual index that represents where an item is in the HashMap
+    real_index: usize,
```

Comment:
It would be better if these were `Cell<usize>`s, to avoid requiring `&mut self` (and thus extra GC write barriers) in  `next/pair_at`.